### ISSUE TITLE
Add parameter name String to ParamsSpec field

### ISSUE THREAD

**Issue by umbreak** [akka/akka-http#3566]
The trait [ParamsSpec](https://github.com/akka/akka-http/blob/master/akka-http/src/main/scala/akka/http/scaladsl/server/directives/ParameterDirectives.scala#L73) could also contain the following:
```scala
def name: String
```
being `name` the requested query parameter field name. This would be very useful for us because we want to override the parameter directive adding some extra action (adding an attribute to the `HttpRequest`):

```scala
  private val atKey: AttributeKey[Set[String]] = AttributeKey("allowed-query-params")

  private def trackQueryParam(name: String): Directive0 =
    optionalAttribute(atKey).flatMap { paramsOpt =>
      mapRequest(addAttribute(atKey, paramsOpt.fold(Set(name))(_ + name)))
    }

def param(pdef1: ParamSpec): Directive[Tuple1[pdef1.Out]] =
    Directives.parameter(pdef1.name) & trackQueryParam(pdef1.name)
```

Since the field name is not currently accessible from `ParamSpec` we have to do some ugly work around to make that work currently.

**Comment by jrudolph** [comment]
Thanks for the suggestion, @umbreak. Even with that addition, wouldn't it still require for you to provide all the overloads of parameters for different numbers of arguments (which we provide via code generation)?

For that reason, my concern is that this feature would not really be generally applicable.

Maybe we can rather discuss you use case on a higher-level to see if it can be provided in some way?

**Comment by umbreak** [comment]
I don't think it is a problem that `ParamsSpec` has the information about which is the query parameter name, since all the calls to `ParamsSpec` must have it.

I'm not interested into supporting all the overrides, since I'm not interested in using `parameters` Directive.

My use case, as you can see in the example is to track the query parameters defined for a certain Route.
We keep track of them using an `AttributeKey` where we store the query parameters' name. And tracking them can be useful to allow this directive:

```scala
  val rejectInvalidParams: Directive0 =
    (parameterMap & optionalAttribute(atKey)).tflatMap { case (allParams, allowedParams) =>
      val invalid = allParams.keySet -- allowedParams.getOrElse(Set.empty[String])
      if (invalid.isEmpty) pass
      else reject(validationRejection(s"query parameters not valid: '${invalid.mkString(", ")}'"))
    }
```

This means that for every Route DSL I have I can make sure that no query parameters, besides the ones I've explicitly defined in the route DSL (using `parameter` directive) are present. And If they are, I reject.

Another way to implement that would have been to keep track of the query params unmatched in the `RequestContext` as you do with the Path.
```scala
trait RequestContext {
...

  /** The unused query parameters of this context.
  val unusedQueryParams: Set[String]
}
```

But It is almost the same, and in both cases there is the need to change something in Akka Http.

**Comment by umbreak** [comment]
@jrudolph the use case is to be able to reject a route if a query parameter that is actually used is provided. This is not for testing purposes, but for production purposes in order to inform the client.

Honestly the addition of `def name: String` to `ParamsSpec` should be more than enough for us. And I don't see any problem with `ParamsSpec` exposing the query parameter name.
